applications/nrf_desktop/src/util/hid_reportq.c

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdint.h>
+#include <zephyr/sys/slist.h>
+#include <zephyr/kernel.h>
+
+#include "hid_reportq.h"
+#include "hid_report_desc.h"
+#include "hid_event.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(hid_reportq, CONFIG_DESKTOP_HID_REPORTQ_LOG_LEVEL);
+
+#define MAX_ENQUEUED_REPORTS CONFIG_DESKTOP_HID_REPORTQ_MAX_ENQUEUED_REPORTS
+
+struct enqueued_report {
+	sys_snode_t node;
+	struct hid_report_event *event;
+};
+
+struct counted_list {
+	sys_slist_t list;
+	size_t node_count;
+};
+
+struct hid_reportq {
+	struct counted_list reports[ARRAY_SIZE(input_reports)];
+	uint16_t enabled_report_idx_bm;
+	uint8_t last_sent_report_idx;
+	uint8_t report_max;
+	uint8_t report_cnt;
+	const void *sub_id;
+};
+
+static struct hid_reportq queues[CONFIG_DESKTOP_HID_REPORTQ_QUEUE_COUNT];
+
+/* Ensure that enabled_report_idx_bm can handle all of the report indexes. */
+BUILD_ASSERT(ARRAY_SIZE(input_reports) <= 16);
+
+static struct enqueued_report *get_enqueued_report(struct counted_list *cnt_list)
+{
+	sys_snode_t *node = sys_slist_get(&cnt_list->list);
+
+	if (!node) {
+		return NULL;
+	}
+
+	__ASSERT_NO_MSG(cnt_list->node_count > 0);
+	cnt_list->node_count--;
+
+	return CONTAINER_OF(node, struct enqueued_report, node);
+}
+
+static void enqueue_report(struct counted_list *cnt_list, struct enqueued_report *report)
+{
+	__ASSERT_NO_MSG(cnt_list->node_count < MAX_ENQUEUED_REPORTS);
+	cnt_list->node_count++;
+
+	sys_slist_append(&cnt_list->list, &report->node);
+}
+
+static struct hid_report_event *get_enqueued_event(struct counted_list *cnt_list)
+{
+	struct enqueued_report *report = get_enqueued_report(cnt_list);
+
+	if (!report) {
+		return NULL;
+	}
+
+	struct hid_report_event *event = report->event;
+
+	k_free(report);
+
+	return event;
+}
+
+static void drop_enqueued_events(struct counted_list *cnt_list)
+{
+	struct hid_report_event *event = get_enqueued_event(cnt_list);
+
+	while (event) {
```

Comment:
consider migration to the `do while` syntax

Reply:
Discussed offline. We will leave it as is.